diffuse: remove unnecessary trigonometry

flannelhead opened the pull request:
The trigonometric calculations in the heat equation calculation are easy to replace by simpler FLOPs. This should make for a nice performance gain at least on the CPU (seems to me this is the case). How to measure it in a reproducible manner? @ralfbrown I recall seeing some performance measurements in your optimization PRs – how did you come up with those?

In separate commits there are also a couple of further optimizations (mainly removal of divisions) but these are not likely as significant.

Test report with the patch applied:
```
Test 0086-diffuse
      Image mire1.cr2
      CPU & GPU version differ by 42566 pixels
      CPU vs. GPU report :
      ----------------------------------
      Max dE                   : 3.29131
      Avg dE                   : 0.00643
      Std dE                   : 0.05739
      ----------------------------------
      Pixels below avg + 0 std : 98.48 %
      Pixels below avg + 1 std : 98.49 %
      Pixels below avg + 3 std : 98.60 %
      Pixels below avg + 6 std : 99.02 %
      Pixels below avg + 9 std : 99.66 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
      Expected CPU vs. current CPU report :
      ----------------------------------
      Max dE                   : 4.14407
      Avg dE                   : 0.01192
      Std dE                   : 0.08298
      ----------------------------------
      Pixels below avg + 0 std : 97.43 %
      Pixels below avg + 1 std : 97.45 %
      Pixels below avg + 3 std : 97.82 %
      Pixels below avg + 6 std : 99.25 %
      Pixels below avg + 9 std : 99.75 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
  OK

Total test 1
Errors     0
```
Without the patch:
```
Test 0086-diffuse
      Image mire1.cr2
      CPU & GPU version differ by 98923 pixels
      CPU vs. GPU report :
      ----------------------------------
      Max dE                   : 4.93617
      Avg dE                   : 0.01611
      Std dE                   : 0.09499
      ----------------------------------
      Pixels below avg + 0 std : 96.46 %
      Pixels below avg + 1 std : 96.54 %
      Pixels below avg + 3 std : 97.24 %
      Pixels below avg + 6 std : 99.23 %
      Pixels below avg + 9 std : 99.81 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
      Expected CPU vs. current CPU report :
      ----------------------------------
      Max dE                   : 1.18492
      Avg dE                   : 0.00057
      Std dE                   : 0.01768
      ----------------------------------
      Pixels below avg + 0 std : 99.88 %
      Pixels below avg + 1 std : 99.88 %
      Pixels below avg + 3 std : 99.88 %
      Pixels below avg + 6 std : 99.88 %
      Pixels below avg + 9 std : 99.88 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
  OK

Total test 1
Errors     0
```
There's some difference caused by this patch but I think the end result is more accurately since we save a roundtrip via `atan -> sin` and `atan -> cos` which are always approximations. @aurelienpierre will have the final word, though.

PS. I'm going to read through the literature for this module and see if there are more optimizations to make (although this was the only low-hanging one I spotted so far). Similar trigonometry optimizations seem to be available for Color balance RGB, but I'll look at them later because the module is already quite fast.

flannelhead commented:
@aurelienpierre a couple of other remarks I noticed when looking at the code (I regret not reading it in the PR phase):

* Is the comment at https://github.com/darktable-org/darktable/blob/05f381e958b2e04e82f9ae8630262f0cfdfd6845/src/iop/diffuse.c#L680 or the code out of date? It seems the Laplacian is calculated as the gradient of `HF` even though the comment says otherwise. The code was changed in https://github.com/darktable-org/darktable/commit/8052ba8053976d1a01adc462adf7ea711ff71695#diff-e7424db92a0af5cdc106e18457769e75e88c3b70bbac59b7bf52568691f8afc0 so I believe the code is correct and the comment should be removed.
* The comment at https://github.com/darktable-org/darktable/blob/05f381e958b2e04e82f9ae8630262f0cfdfd6845/src/iop/diffuse.c#L484 says that y (partial derivative w.r.t. y?) is at index 0 and x at index 1. Are the arguments of the `atan2()` at https://github.com/darktable-org/darktable/blob/05f381e958b2e04e82f9ae8630262f0cfdfd6845/src/iop/diffuse.c#L688 the right way around? It is not immediately obvious even though I skimmed the literature briefly.

ralfbrown commented:
For  performance numbers I typically use `-d  perf` and average 2-3 runs of darktable-cli in each condition.  For finer-grained timings, I'll surround the code in question with
```
double start = dt_get_wtime();
  ....
fprintf(stderr,"X took %g sec\n",dt_get_wtime() - start);
```

aurelienpierre commented:
> PS. I'm going to read through the literature for this module and see if there are more optimizations to make (although this was the only low-hanging one I spotted so far). Similar trigonometry optimizations seem to be available for Color balance RGB, but I'll look at them later because the module is already quite fast.

You can't do similar optimisations in color balance because `atan2` returns hue and we need it explicitly. What you did here is clever though, since we don't explicitly need angles. Your review is very much appreciated.

> The code was changed in 8052ba8#diff-e7424db92a0af5cdc106e18457769e75e88c3b70bbac59b7bf52568691f8afc0 so I believe the code is correct and the comment should be removed.

Yes, comment is outdated.

Regarding the orientation of dx / dy, 
```
static inline void find_gradient(const float pixels[9][4], const size_t c, float xy[2])
{
  // Compute the gradient with centered finite differences in a 3×3 stencil
  // warning : x is vertical, y is horizontal
  xy[0] = (pixels[7][c] - pixels[1][c]) / 2.f; // dx
  xy[1] = (pixels[5][c] - pixels[3][c]) / 2.f; // dy
}
```
the first element of `xy` is dx (vertical), the second element is dy (horizontal), so 

```
const float theta_grad = atan2f(gradient[1], gradient[0]); 
```

is the correct orientation of the gradient (`atan2(dy, dx)`), unless I missed something in-between (I checked it empirically against synthetic closed shapes, it doesn't propagate gradient out of bounds).

flannelhead commented:
Thanks for the review! I'll follow up with performance measurements and check each individual commit's impact before this one is merged. Is that ok?

aurelienpierre commented:
Yup, that's great. Thanks ! Never trust my code too much, I know I don't.

ralfbrown commented:
@aurelienpierre  In looking over the code, I see that you store all of the decomposed wavelet scales to then apply `heat_PDE_diffusion` on them from coarsest to finest.  Is there any possibility of reversing the processing order for `heat_PDE_diffusion` and accumulating the per-scale results in the final output buffer?  That would eliminate the need to store all of the scales and cut memory requirements by more than half (it's what I did for contrast equalizer).  If I'm  looking at things correctly, the `tiling_factor` corresponding to the memory requirement would go from 16.25 to 5.25 or 6.25, depending on whether the reversal introduces the need for an additional temporary buffer.

aurelienpierre commented:
> Is there any possibility of reversing the processing order for heat_PDE_diffusion and accumulating the per-scale results in the final output buffer?

No. Notice it's what I did in filmic reconstruction: we process each scale in parallel. But here we need to process sequentially such that each layer LF (residual) is inited with the coarser scales. This ensures faster convergence of the multi-scale PDE scheme (the first implementation used the same parallel implementation as filmic's one, but it needed a lot more iterations).

flannelhead commented:
Here are the perf test results on Intel i7-6600U, HD Graphics 520. Execution times for the diffuse module on CPU and GPU, average of two runs.

```
Version   Avg CPU   Perf CPU   Avg GPU   Perf GPU
master      19.08       1.00      2.30       1.00
patch 1     10.59       1.80      2.30       1.00
patch 2     10.68       1.79      2.32       0.99
patch 3     10.39       1.84      2.30       1.00
patch 4     10.32       1.85      2.27       1.01
```

The first commit brings the greatest performance wins for CPU, GPU perf is basically intact. The CPU result is as expected, the removal of trigonometric functions brings the bulk of the improvement. I'm mildly surprised that the GPU performance didn't get better at all. I'm thinking of dropping at least the second patch (combining the `sin(theta) * cos(theta)` product into a single argument) – probably inlining is pretty effective there and the compiler does better.

The other two patches seem to probably yield a little improvement, so they might as well be merged. What do you think?

flannelhead commented:
Another thought: there are still four evaluations of `expf()` in the same place in the code. They are probably quite expensive at least on the CPU. I can't think of any easy way to directly replace them with anything faster, but what about approximating them somehow, for example a Taylor series? While doing that, one could also probably get rid of the `hypotf()` calls as well and use the squared magnitudes instead (since squared magnitudes are all we need for the sin and cos calculations, since actually products of the two are only ever used).

ralfbrown commented:
There's a `dt_fast_expf` in common/math.h.  It has a worst-case relative error of several percent.

aurelienpierre commented:
Isn't `expf` supposed to have a SIMD variant ?

flannelhead commented:
Would probably have to look at the generated code to be sure. But it seems using `dt_fast_exp()` there yields yet another speedup of almost three seconds on the CPU. The output image will be different though. I'll try the Taylor approach tomorrow.

aurelienpierre commented:
>

flannelhead commented:
Yes but one could do the Taylor expansion of `expf(-k * sqrtf(x))` about `x` at a suitable point, say `x = 1`.

aurelienpierre commented:
Well, if you are that motivated… ;-)

aurelienpierre commented:
Ok, I did a test : https://godbolt.org/z/7fxKjznsW

```
void c1(float magnitude[4], float anisotropy[4], float c2[4])
{
    c2[0] = expf(-magnitude[0] / anisotropy[0]);
    c2[1] = expf(-magnitude[1] / anisotropy[1]);
    c2[2] = expf(-magnitude[2] / anisotropy[2]);
    c2[3] = expf(-magnitude[3] / anisotropy[3]);
} 
```
compiles as

```
c1:
  push r12
  vmovss xmm0, DWORD PTR [rdi]
  mov r12, rdi
  push rbp
  vxorps xmm0, xmm0, XMMWORD PTR .LC0[rip]
  mov rbp, rsi
  push rbx
  vdivss xmm0, xmm0, DWORD PTR [rsi]
  mov rbx, rdx
  call expf
  vmovss DWORD PTR [rbx], xmm0
  vmovss xmm0, DWORD PTR [r12+4]
  vxorps xmm0, xmm0, XMMWORD PTR .LC0[rip]
  vdivss xmm0, xmm0, DWORD PTR [rbp+4]
  call expf
  vmovss DWORD PTR [rbx+4], xmm0
  vmovss xmm0, DWORD PTR [r12+8]
  vxorps xmm0, xmm0, XMMWORD PTR .LC0[rip]
  vdivss xmm0, xmm0, DWORD PTR [rbp+8]
  call expf
  vmovss DWORD PTR [rbx+8], xmm0
  vmovss xmm0, DWORD PTR [r12+12]
  vxorps xmm0, xmm0, XMMWORD PTR .LC0[rip]
  vdivss xmm0, xmm0, DWORD PTR [rbp+12]
  call expf
  vmovss DWORD PTR [rbx+12], xmm0
  pop rbx
  pop rbp
  pop r12
  ret
```

but 
```
void c0(float magnitude[4], float anisotropy[4], float c2[4])
{
    anisotropy = (float*)__builtin_assume_aligned(anisotropy, 64);
    c2 = (float*)__builtin_assume_aligned(c2, 64);
    magnitude = (float*)__builtin_assume_aligned(magnitude, 64);

    for(int c = 0; c < 4; c++) c2[c] = expf(-magnitude[c] / anisotropy[c]);
} 
```

produces

```
c0:
  push r13
  mov r13, rdi
  push r12
  mov r12, rsi
  push rbp
  mov rbp, rdx
  push rbx
  xor ebx, ebx
  sub rsp, 8
.L35:
  vmovss xmm0, DWORD PTR [r13+0+rbx]
  vxorps xmm0, xmm0, XMMWORD PTR .LC0[rip]
  vdivss xmm0, xmm0, DWORD PTR [r12+rbx]
  call expf
  vmovss DWORD PTR [rbp+0+rbx], xmm0
  add rbx, 4
  cmp rbx, 16
  jne .L35
  add rsp, 8
  pop rbx
  pop rbp
  pop r12
  pop r13
  ret
```

So it needs an explicit loop to vectorize, it seems.

aurelienpierre commented:
On the other hand, 

```
static inline float dt_fast_expf(const float x)
{
  // meant for the range [-100.0f, 0.0f]. largest error ~ -0.06 at 0.0f.
  // will get _a_lot_ worse for x > 0.0f (9000 at 10.0f)..
  const int i1 = 0x3f800000u;
  // e^x, the comment would be 2^x
  const int i2 = 0x402DF854u; // 0x40000000u;
  // const int k = CLAMPS(i1 + x * (i2 - i1), 0x0u, 0x7fffffffu);
  // without max clamping (doesn't work for large x, but is faster):
  const int k0 = i1 + x * (i2 - i1);
  union {
      float f;
      int k;
  } u;
  u.k = k0 > 0 ? k0 : 0;
  return u.f;
}

void c0(float magnitude[4], float anisotropy[4], float c2[4])
{
    for(int c = 0; c < 4; c++)
        c2[c] = dt_fast_expf(-magnitude[c] / anisotropy[c]);
} 
```

compiles to something really shitty:
```
c0:
  lea rcx, [rsi+4]
  mov rax, rdx
  sub rax, rcx
  cmp rax, 8
  jbe .L35
  lea rcx, [rdi+4]
  mov rax, rdx
  sub rax, rcx
  cmp rax, 8
  jbe .L35
  vmovups xmm6, XMMWORD PTR [rdi]
  vpxor xmm1, xmm1, xmm1
  vxorps xmm0, xmm6, XMMWORD PTR .LC0[rip]
  vdivps xmm0, xmm0, XMMWORD PTR [rsi]
  vmulps xmm0, xmm0, XMMWORD PTR .LC1[rip]
  vaddps xmm0, xmm0, XMMWORD PTR .LC2[rip]
  vcvttps2dq xmm0, xmm0
  vpmaxsd xmm0, xmm0, xmm1
  vmovups XMMWORD PTR [rdx], xmm0
  ret
.L35:
  vmovss xmm1, DWORD PTR [rdi]
  vmovss xmm4, DWORD PTR .LC4[rip]
  vmovss xmm5, DWORD PTR .LC3[rip]
  vxorps xmm1, xmm1, xmm5
  vdivss xmm1, xmm1, DWORD PTR [rsi]
  vmulss xmm1, xmm1, xmm4
  vmovss xmm3, DWORD PTR .LC5[rip]
  vpxor xmm2, xmm2, xmm2
  vaddss xmm1, xmm1, xmm3
  vcvttss2si eax, xmm1
  vmovd xmm0, eax
  vpmaxsd xmm0, xmm0, xmm2
  vmovss DWORD PTR [rdx], xmm0
  vmovss xmm1, DWORD PTR [rdi+4]
  vxorps xmm1, xmm1, xmm5
  vdivss xmm1, xmm1, DWORD PTR [rsi+4]
  vmulss xmm1, xmm1, xmm4
  vaddss xmm1, xmm1, xmm3
  vcvttss2si eax, xmm1
  vpinsrd xmm0, xmm2, eax, 0
  vpmaxsd xmm0, xmm0, xmm2
  vmovss DWORD PTR [rdx+4], xmm0
  vmovss xmm1, DWORD PTR [rdi+8]
  vxorps xmm1, xmm1, xmm5
  vdivss xmm1, xmm1, DWORD PTR [rsi+8]
  vmulss xmm1, xmm1, xmm4
  vaddss xmm1, xmm1, xmm3
  vcvttss2si eax, xmm1
  vpinsrd xmm0, xmm2, eax, 0
  vpmaxsd xmm0, xmm0, xmm2
  vmovss DWORD PTR [rdx+8], xmm0
  vmovss xmm0, DWORD PTR [rdi+12]
  vxorps xmm0, xmm0, xmm5
  vdivss xmm0, xmm0, DWORD PTR [rsi+12]
  vmulss xmm0, xmm0, xmm4
  vaddss xmm0, xmm0, xmm3
  vcvttss2si eax, xmm0
  vpinsrd xmm0, xmm2, eax, 0
  vpmaxsd xmm0, xmm0, xmm2
  vmovss DWORD PTR [rdx+12], xmm0
  ret
```

ralfbrown commented:
Your second example isn't vectorized either, it's just looping four times over the single call to the scalar `expf`.  Other than the vxorps, all of the floating-point instructions are scalar ('ss' suffix instead of 'ps').  It seems that the compiler uses vxorps to zero a vector register whether it plans to use the entire register or just its lowest element.

aurelienpierre commented:
Right, the code looked prettier, I saw `expf` written just once, and I don't really read x86 assembly.

aurelienpierre commented:
> Yes but one could do the Taylor expansion of expf(-k * sqrtf(x)) w.r.t. x at a suitable point, say x = 1.

Thinking about it, you still need the sqrt to compute sin and cos from the hypotenuse.

flannelhead commented:
> Thinking about it, you still need the sqrt to compute sin and cos from the hypotenuse.
That was not an issue, you see the sin and cos are always used as products of each other so the hypotenuse squared can be used in the calculation of the products.

However, I couldn't expand `expf(k * sqrtf(x))` – the derivative of the square root does not exist at `x = 0` and thus it's not feasible to approximate this composite function near zero by a Taylor polynomial.

I approximated `expf(k * x)` instead and left the square roots as they are. The result is a noticeable performance gain at the cost of accuracy.

```
Test 0086-diffuse
      Image mire1.cr2
      CPU & GPU version differ by 42720 pixels
      CPU vs. GPU report :
      ----------------------------------
      Max dE                   : 3.29131
      Avg dE                   : 0.00642
      Std dE                   : 0.05721
      ----------------------------------
      Pixels below avg + 0 std : 98.47 %
      Pixels below avg + 1 std : 98.48 %
      Pixels below avg + 3 std : 98.60 %
      Pixels below avg + 6 std : 99.02 %
      Pixels below avg + 9 std : 99.66 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
      Expected CPU vs. current CPU report :
      ----------------------------------
      Max dE                   : 18.88941
      Avg dE                   : 0.16594
      Std dE                   : 0.42714
      ----------------------------------
      Pixels below avg + 0 std : 75.94 %
      Pixels below avg + 1 std : 90.99 %
      Pixels below avg + 3 std : 98.22 %
      Pixels below avg + 6 std : 99.58 %
      Pixels below avg + 9 std : 99.86 %
      ----------------------------------
      Pixels above tolerance   : 0.64 %
```

Then I tried just replacing the `expf()` with `dt_fast_expf()`. The results are more accurate and the performance gain is also very nice.

```
Test 0086-diffuse
      Image mire1.cr2
      CPU & GPU version differ by 43225 pixels
      CPU vs. GPU report :
      ----------------------------------
      Max dE                   : 2.71703
      Avg dE                   : 0.00650
      Std dE                   : 0.05757
      ----------------------------------
      Pixels below avg + 0 std : 98.45 %
      Pixels below avg + 1 std : 98.47 %
      Pixels below avg + 3 std : 98.58 %
      Pixels below avg + 6 std : 99.01 %
      Pixels below avg + 9 std : 99.66 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
      Expected CPU vs. current CPU report :
      ----------------------------------
      Max dE                   : 5.03928
      Avg dE                   : 0.19372
      Std dE                   : 0.28851
      ----------------------------------
      Pixels below avg + 0 std : 63.48 %
      Pixels below avg + 1 std : 82.78 %
      Pixels below avg + 3 std : 98.92 %
      Pixels below avg + 6 std : 99.94 %
      Pixels below avg + 9 std : 100.00 %
      ----------------------------------
      Pixels above tolerance   : 0.02 %
```

Here are the performance timings:

```
Version                      Avg CPU   Perf CPU   Avg GPU   Perf GPU
master                         19.75       1.00      2.30       1.00
trigonometry optimizations      9.09       2.17      2.25       1.02
trigonometry + dt_fast_expf     6.72       2.94      2.27       1.02
```

So these optimizations total to a nearly 3x speedup on CPU. GPU doesn't show any tangible benefit, sadly.

I'd consider this PR ready for (another) review on my part. Any further optimizations should probably guided by profiling.

flannelhead commented:
A side note: in my tests, the preview pixel pipe still seems to be usually executed on the CPU. This slows down the GUI a lot (`working...` is displayed on the image for a long time while the code runs on the CPU). As we know this module is usually quite slow, is there any way to always force it to run on the GPU if one is available?

flannelhead commented:
@aurelienpierre another question that arose from reading the code: why are the matrix elements clamped [here](https://github.com/darktable-org/darktable/blob/950232e0a7702c3dc193bfaf8e282836dc10189f/src/iop/diffuse.c#L532-L534)? The diagonal ones might have a negative sign related to the direction of the gradient or the isophote, but clamping to zero destroys that information.

aurelienpierre commented:
> Clamping of the diagonal terms is not that useful either since by definition sin(theta)^2 + cos(theta)^2 is between 0 and 1. Adding the multiplier c2, also in range 0 to 1, to either term doesn't bring the result out of range.

Along the design process, I met some issues with c2 being out of float precision range and I did that for safety, but maybe it's safe to remove it now.

> Clamping of the diagonal terms is not that useful either since by definition sin(theta)^2 + cos(theta)^2 is between 0 and 1. Adding the multiplier c2, also in range 0 to 1, to either term doesn't bring the result out of range.

Right, that's a mistake.

> So these optimizations total to a nearly 3x speedup on CPU. GPU version doesn't show any tangible performance benefit, sadly.

GPU is good at hiding suboptimal code, but I would stick to the `native_exp` instead of the dt variant since it should vectorize.

I'll review tomorrow.

flannelhead commented:
> GPU is good at hiding suboptimal code, but I would stick to the native_exp instead of the dt variant since it should vectorize.

Alright, updated now so that only CPU path uses `dt_fast_expf()` and GPU path uses `native_exp()`. Makes sense to me, although there will be a slight (probably not noticeable) difference between the output of the two versions.

I also removed the clamping of matrix elements and added some necessary clamping to the gradient magnitude calculation to avoid division by zero. Hopefully this also removes some of the issues with floating point precision you were facing. At least the calculation of the sines and cosines should be fairly stable now.

aurelienpierre commented on data/kernels/diffuse.cl:
```diff
@@ -254,34 +258,34 @@ diffuse_pde(read_only image2d_t HF, read_only image2d_t LF,
-    const float4 sin_theta_grad_sq = sqf(sin_theta_grad);
-    const float4 cos_theta_lapl_sq = sqf(cos_theta_lapl);
-    const float4 sin_theta_lapl_sq = sqf(sin_theta_lapl);
+    const float4 magnitude_grad = native_sqrt(sqf(gradient[0]) + sqf(gradient[1]));
```
this is exactly equivalent to `hypotf`, why change it ?

aurelienpierre commented on data/kernels/diffuse.cl:
```diff
@@ -254,34 +258,34 @@ diffuse_pde(read_only image2d_t HF, read_only image2d_t LF,
-    const float4 cos_theta_lapl_sq = sqf(cos_theta_lapl);
-    const float4 sin_theta_lapl_sq = sqf(sin_theta_lapl);
+    const float4 magnitude_grad = native_sqrt(sqf(gradient[0]) + sqf(gradient[1]));
+    const float4 norm_grad = 1.f / fmax(magnitude_grad, NORM_MIN);
```
This one took me a while to figure out. 

What you compute here is not the norm, but the hypotenuse, and you divide `grad_x` and `grad_y` by the hypotenuse to get the cos and sin of `arg(grad)`, not to normalize it. Although arithmetically equivalent, the concept is different and it's not immediately clear that `gradient[0]` becomes `cos(arg(gradient))` at the next step. 

I'm not nitpicking, there are sisters methods using Total Variation regularization that also rely on normalizing the laplacian by the magnitude of the gradient, and that look just like that. The difference is here, the magnitude of the gradient is used to compute the angle of the rotation matrix, not to regularize the laplacian, so let's make it clear.

I think when `grad = (0, 0)` we should just force `arg(grad) = 0` by convention and therefore, `cos(0) = 1` and `sin(0) = 1`. In that case, `c2 == 1` exactly. Thresholding by NORM_MIN is inaccurate in that regard.

I would rewrite it as follow (including comments):

```C
const float4 magnitude_grad = hypotf(gradient[0],  gradient[1]);

// Compute cos(arg(grad)) = x / hypot - force arg(grad) = 0 if hypot == 0
gradient[0] = (magnitude_grad != 0.f) ? gradient[0] / magnitude_grad
                                                              : 1.f; // cos(0)
// Compute sin (arg(grad))= y / hypot - force arg(grad) = 0 if hypot == 0
gradient[1] = (magnitude_grad != 0.f) ? gradient[1] / magnitude_grad
                                                              : 0.f; // sin(0)

// Warning : now gradient[2] = { cos(arg(grad)) , sin(arg(arg)) }
```

aurelienpierre reviewed (changes requested):
Almost good to go

aurelienpierre reviewed (approved):
Awesome, thanks !
